### PR TITLE
fix: clip-onnx spec 9 & 11

### PR DESCRIPTION
**Pull request by ganler**
clip-v6 and clip-v9 only support float(16/32/64) as inputs such that prior implementation will make exported `clamp(int_min, int_max)` fail in ONNXRuntime. 

image

To fix it, use the "cast-clip-cast" pattern as a workaround.

cc: @peterbell10 @BowenBao

### CODE REVIEW

**Comment by ganler** [comment]
@thiagocrepaldi @BowenBao Could you kindly review this PR? :-)

**Review by thiagocrepaldi — COMMENTED** [review]
Why doesn't Operators.md list opset9 for clip? I expected to either have opset 9 listed or not having a symbolic for clip on symbolic_opset9.py.

@ganler Please clarify

**Comment by ganler** [comment]
@thiagocrepaldi Thanks for the reply! and I am sorry for the lack of clarification.

You are right that Clip does not have a version called version 9. But I just found that the clip operator was originally implemented in `symbolic_opset9` which I thought represents new/changed operators in version 9.

According to the [doc](https://github.com/onnx/onnx/blob/main/docs/Operators.md), Clip has version [13, 12, 11, 6, 1] where [11, 6, 1] do not support integer-typed clip but PyTorch seems to export that unsupported operator still, leading to an error in model importation for ONNXRuntime. 

Therefore, I was thinking maybe we can still support that by apply the `cast-clip-cast` pattern.

Regarding the versions, you are right that maybe we should not put the clip implementation in `symbolic_opset9.py` as clip is not updated in opset 9. Could you kindly indicate whether we should have a new file `symbolic_opset6.py` or put it elsewhere? Thanks!

**Comment by thiagocrepaldi** [comment]
> @thiagocrepaldi Thanks for the reply! and I am sorry for the lack of clarification.
> 
> You are right that Clip does not have a version called version 9. But I just found that the clip operator was originally implemented in `symbolic_opset9` (
> 
> https://github.com/pytorch/pytorch/blob/4b47047dae11ab764dfa6f2ff5ff8342b2161361/torch/onnx/symbolic_opset9.py#L1594
> 
> ) which I thought represents new/changed operators in version 9.
> According to the [doc](https://github.com/onnx/onnx/blob/main/docs/Operators.md), Clip has version [13, 12, 11, 6, 1] where [11, 6, 1] do not support integer-typed clip but PyTorch seems to export that unsupported operator still, leading to an error in model importation for ONNXRuntime.
> 
> Therefore, I was thinking maybe we can still support that by apply the `cast-clip-cast` pattern.
> 
> Regarding the versions, you are right that maybe we should not put the clip implementation in `symbolic_opset9.py` as clip is not updated in opset 9. Could you kindly indicate whether we should have a new file `symbolic_opset6.py` or put it elsewhere? Thanks!

Given the spec, does the behavior implemented on symbolic_opset9.py belongs to 1, 6, 11, 12 or 13?

Whatever is exported by pytorch has to fit the onnx spec. If that is not the case, then we have a bug on pytorch as opposed to the onnx symbolic

**Comment by ganler** [comment]
> Given the spec, does the behavior implemented on symbolic_opset9.py belongs to 1, 6, 11, 12 or 13?

@thiagocrepaldi 

IMHO, Clip in `symbolic_opset9.py` intends to implement Clip opset [6](https://github.com/onnx/onnx/blob/main/docs/Changelog.md#Clip-6) because in opset 6, "min" and "max" are floats. And Clip in `symbolic_opset9.py` expects "min" and "max" to be "f" in `_parse_arg`. But Clip opset < 12 (including opset 6) does not support `int` which requires a trick like `cast-float-clip-cast` to implement that.

https://github.com/pytorch/pytorch/blob/4b47047dae11ab764dfa6f2ff5ff8342b2161361/torch/onnx/symbolic_opset9.py#L1611

> Whatever is exported by pytorch has to fit the onnx spec. If that is not the case, then we have a bug on pytorch as opposed to the onnx symbolic.

Agreed. So I would suggest:
1. Support `clip<int>` in opset < 12 using `cast-float-clip-cast` pattern;
2. Move the implementation to a new file called `symbolic_opset6.py` or so;

File: torch/onnx/symbolic_opset9.py [hunk 1]
```diff
@@ -1591,6 +1591,30 @@ def pow(g, self, exponent):
...
+
+    # clip-6 is only available to float, double, and float16.
+    # for other types, we first cast to float, then do clip and cast back.
+    origin_dtype = self.type().scalarType()
+    is_fp_clamp = self.type().scalarType(
+    ) == "Float" or self.type().scalarType() == "Double"
+
+    if not is_fp_clamp:
+        warnings.warn(
+            "ONNX opset < 11 does not support clamp/clip to non-float types. Using cast-clip-cast instead.")
```
Comment (thiagocrepaldi) [inline torch/onnx/symbolic_opset9.py:1606]: It does not make sense reference a future opset in this message, right?

Comment (thiagocrepaldi) [reply torch/onnx/symbolic_opset9.py:1606]: If this implementation belongs to opset 6, move this piece of code there.

**Comment by thiagocrepaldi** [comment]
> > Given the spec, does the behavior implemented on symbolic_opset9.py belongs to 1, 6, 11, 12 or 13?
> 
> @thiagocrepaldi
> 
> IMHO, Clip in `symbolic_opset9.py` intends to implement Clip opset [6](https://github.com/onnx/onnx/blob/main/docs/Changelog.md#Clip-6) because in opset 6, "min" and "max" are floats (and in later versions, say opset [11](https://github.com/onnx/onnx/blob/main/docs/Changelog.md#clip-11), its parameter type becomes "T" which is the data type of input tensors). And Clip in `symbolic_opset9.py` expects "min" and "max" to be "f" in `_parse_arg`. But Clip opset < 12 (including opset 6) does not support `int` which requires a trick like `cast-float-clip-cast` to implement that.
> 
> https://github.com/pytorch/pytorch/blob/4b47047dae11ab764dfa6f2ff5ff8342b2161361/torch/onnx/symbolic_opset9.py#L1611
> 
> > Whatever is exported by pytorch has to fit the onnx spec. If that is not the case, then we have a bug on pytorch as opposed to the onnx symbolic.
> 
> Agreed. So I would suggest:
> 
> 1. Support `clip<int>` in opset < 12 using `cast-float-clip-cast` pattern;
> 2. Move the implementation to a new file called `symbolic_opset6.py` or so;
> 
> I am happy to implement that in this PR if it looks to you. :-)

Looks sensible to me. let me check with a colleague

@BowenBao how do you feel about moving clip-9 code to clip-6? Looking the onnx spec, there is no such clip-9 and clip-6 does support float for min/max

File: torch/onnx/symbolic_opset9.py (hunk 1 above)
Comment (thiagocrepaldi) [reply torch/onnx/symbolic_opset9.py:1606]: On a second note, there is no `symbolic_opset6.py` in the repo. Maybe we should leave it on symbolic 9 and check with pytorch folks what is going on.

File: torch/onnx/symbolic_opset11.py [hunk 2]
```diff
@@ -17,6 +17,30 @@
 
 # This file exports ONNX ops for opset 11
 
+def _clamp_common_types(g, self, min, max):
+    # clip-6 is only available to float, double, and float16.
```
Comment (thiagocrepaldi) [inline torch/onnx/symbolic_opset11.py:21]:
```suggestion
    # clip-11 is only available to float, double, and float16.
```

File: torch/onnx/symbolic_opset11.py [hunk 3]
```diff
@@ -17,6 +17,30 @@
...
+    # for other types, we first cast to float, then do clip and cast back.
+    origin_dtype = self.type().scalarType()
+    is_fp_clamp = self.type().scalarType(
+    ) == "Float" or self.type().scalarType() == "Double"
+
+    if not is_fp_clamp:
+        warnings.warn(
+            "ONNX opset < 11 does not support clamp/clip to non-float types. Using cast-clip-cast instead.")
```
Comment (thiagocrepaldi) [inline torch/onnx/symbolic_opset11.py:29]:
```suggestion
            "ONNX opset 11 does not support clamp/clip to non-float types. Using cast-clip-cast instead.")
```

File: torch/onnx/symbolic_opset9.py (hunk 1 above)
Comment (thiagocrepaldi) [inline torch/onnx/symbolic_opset9.py:1606]:
```suggestion
            "ONNX opset 6 does not support clamp/clip to non-float types. Using cast-clip-cast instead.")
```

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 4]
```diff
@@ -3421,6 +3421,18 @@ def forward(self, x):
         x = torch.randn(3, 3)
         self.run_test(MyReluModule(), x)
 
+    @skipIfUnsupportedMinOpsetVersion(9)
+    def test_clip9(self):
+        class MyClip9(torch.nn.Module):
```
Comment (thiagocrepaldi) [inline test/onnx/test_pytorch_onnx_onnxruntime.py:3426]:
```suggestion
        class MyClip6(torch.nn.Module):
```

Comment (thiagocrepaldi) [reply test/onnx/test_pytorch_onnx_onnxruntime.py:3426]: although you can leave the clip-6 inside opset9 file, name it as the correct name to let documented the intended onnx spec

**Comment by ganler** [comment]
@thiagocrepaldi Thanks for the review. Comments should be resolved! Could you also take a look at https://github.com/pytorch/pytorch/pull/70571 ?

**Review by thiagocrepaldi — APPROVED** [review]

**Comment by AllenTiTaiWang** [comment]
cc @thiagocrepaldi

**Review by BowenBao — APPROVED** [review]

**Comment by thiagocrepaldi** [comment]
@ganler do you mind rebasing this?

**Pull request closed without merging**